firedrake/matrix_free/formmanipulation.py

```diff
@@ -0,0 +1,82 @@
+from __future__ import absolute_import
+
+import numpy
+
+from ufl import as_vector
+from ufl.classes import Zero
+from ufl.algorithms.map_integrands import map_integrand_dags
+from ufl.corealg.map_dag import MultiFunction
+
+from firedrake.ufl_expr import Argument
+
+
+class ExtractSubBlock(MultiFunction):
```

Comment:
It's not quite.  I think one might be able to avoid code duplication.  The difference is this.  The form splitter splits all components of mixed spaces into separate pieces.  For this one, you can specify which sub-blocks of the mixed space should be split out.  Therefore imagine a matrix:

```
A B C
D E F
G H I
```

The form splitter will give you 9 sub-forms, one for each of A-I.  With ExtractSubBlock you pass in the space indices that you want.  E.g.:

```python
ExtractSubBlock().split(test_indices=(0, 2), trial_indices=(0, 2))
```
Gives you the form for:

```
A C
G I
```

This is used for recursive splitting of forms for fieldsplit.